source/pcic/client.py

```diff
@@ -27,15 +29,20 @@ def recv(self, number_bytes):
         :param number_bytes: (int) length of bytes
         :return: the data as bytearray
         """
-        data = bytearray()
-        while len(data) < number_bytes:
-            data_part = self.pcicSocket.recv(number_bytes - len(data))
-            if len(data_part) == 0:
-                raise RuntimeError("Connection to server closed")
-            data = data + data_part
-        self.recv_counter += number_bytes
-        return data
-
+        sessionstatus = self.rpcdevice.getParameter("OperatingMode")
+        if sessionstatus == "0":
+            data = bytearray()
+            while len(data) < number_bytes:
+                data_part = self.pcicSocket.recv(number_bytes - len(data))
+                if len(data_part) == 0:
+                    raise RuntimeError("Connection to server closed")
+                data = data + data_part
+            self.recv_counter += number_bytes
+            return data
+
+        else:
+            raise RuntimeError("Session is already open!")
```

Comment:
There are 2 additional OperatingModes. Maybe raise RuntimeErrors with different messages depending on the Operating Mode. like: 

if sessionstatus == 1:
raise RuntimeError("Sensor is in Parametrization Mode. Please close the ifmVisionAssistant and retry.")
elif .....

-1 Booting Mode @cfreundl is this right?
1 Parametrization Mode

Reply:
@Kottanidisan see comment

Reply:
Regarding the possible `OperatingMode` values: it is "0" (run mode), "1" (edit mode) or "2" (simulation mode). Simulation mode is only used internal for testing where you can feed images from the outside into the camera for processing similar to the snapshot processing in edit mode. Otherwise the device behaves just like it does in run mode, i.e. this would be a valid operating mode for PCIC communication.

I see two problems here: first, the ability to communicate with the device over PCIC is independent of whether there is a session opened and also whether the device has been put in edit mode. That's why I have doubts whether this method is a good place to perform this check. I think issue #7 should clarify the uses cases where an error message like this is helpful or even required (this is more a task for @Galoshi), maybe we can find a better location for this check then.

And as we are talking about the location, the impact of this check on the actual reception of PCIC output is huge. With every `recv()` call you now have to perform a full XMLRPC call first, i.e. one complete network roundtrip including an HTTP request. The latency that is introduced here might introduce performance issues if there is a continuously running application with a low evaluation time running on the device.

Reply:
After rereading the feature request in issue #7 I see that it already states that the error message should be returned when "an attempt is made to open a device", so this already sounds like the check is probably more suited in the `__init__` method.